src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfo.cs

```diff
@@ -505,9 +494,9 @@ private static bool PropertyIsVirtual(PropertyInfo? propertyInfo)
             return propertyInfo != null && (propertyInfo.GetMethod?.IsVirtual == true || propertyInfo.SetMethod?.IsVirtual == true);
         }
 
-        private bool DetermineExtensionDataProperty(Dictionary<string, JsonPropertyInfo> cache)
+        private bool DetermineExtensionDataProperty()
```

Comment:
```suggestion
        private bool TryGetExtensionDataProperty([NotNullWhen(true)] out JsonPropertyInfo? extensionDataProperty)
```

and set `DataExtensionProperty` at the caller.